### PR TITLE
Send notification for PGP keys in secret store file

### PR DESCRIPTION
**Pull request by patrickxb**
Also, provide an RPC/CLI command to dismiss the notification on the current device in the future.

### CODE REVIEW

File: go/client/cmd_dismiss.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+func (c *CmdDismiss) GetUsage() libkb.Usage {
+	return libkb.Usage{
+		Config: true,
+	}
+}
+
+func (c *CmdDismiss) Run() error {
+	switch c.name {
+	case PGPStorage:
+		cli, err := GetPGPClient()
```
Comment (maxtaco) [inline go/client/cmd_dismiss.go]: can you fix this to use context?

Comment (patrickxb) [inline go/client/cmd_dismiss.go]: yes, oversight.

File: go/libkb/db.go [hunk 2]
```diff
@@ -139,3 +140,11 @@ const (
 func DbKeyUID(t ObjType, uid keybase1.UID) DbKey {
 	return DbKey{Typ: t, Key: uid.String()}
 }
+
+func DbKeyNotificationDismiss(prefix string, username NormalizedUsername) DbKey {
+	return DbKey{
+		Typ: DBNotificationDismiss,
+		Key: fmt.Sprintf("%s|%s", prefix, username),
```
Comment (maxtaco) [inline go/libkb/db.go]: Use `:`?

File: go/libkb/secret_store_other.go [hunk 3]
```diff
@@ -6,5 +6,77 @@
...
+	return s
 }
+
+func notifySecretStoreCreate(g *GlobalContext, username NormalizedUsername) {
+	g.Log.Debug("got secret store file notifyCreate")
+
+	// check leveldb for existence of notification dismissal
+	dbobj, found, err := g.LocalDb.GetRaw(DbKeyNotificationDismiss(NotificationDismissPGPPrefix, username))
+	if err != nil {
+		g.Log.Debug("notifySecretStoreCreate:  localDb.GetRaw error: %s", err)
```
Comment (maxtaco) [inline go/libkb/secret_store_other.go]: one space

File: go/libkb/secret_store_other.go [hunk 4]
```diff
@@ -6,5 +6,77 @@
...
+		return
+	}
+	if found && string(dbobj) == NotificationDismissPGPValue {
+		g.Log.Debug("notifySecretStoreCreate:  %s already dismissed", NotificationDismissPGPPrefix)
```
Comment (maxtaco) [inline go/libkb/secret_store_other.go]: one space

File: go/libkb/secret_store_other.go [hunk 5]
```diff
@@ -6,5 +6,77 @@
...
+		return
+	}
+	blocks, err := kr.AllPGPBlocks()
+	if err != nil {
+		g.Log.Debug("keyring.AllPGPBlocks error: %s", err)
+		return
+	}
+
+	if len(blocks) == 0 {
+		g.Log.Debug("notifySecretStoreCreate:  no pgp blocks in keyring")
```
Comment (maxtaco) [inline go/libkb/secret_store_other.go]: same

**Comment by patrickxb** [comment]
@keybase/react-hackers need flow help

```[merging inference] \../shared/util/set-notifications.js:38
 38:       notifyCtlSetNotificationsRpc({
           ^ function call
2806:   channels: NotificationChannels
                  ^^^^^^^^^^^^^^^^^^^^ property `pgp`. Property not found in. See: ../shared/constants/types/flow-types.js:2806
 39:         param: {channels: toSend},
                               ^^^^^^ object literal```

i did `make clean && make` in protocol.

**Comment by MarcoPolo** [comment]
on it

File: go/libkb/secret_store_file.go [hunk 6]
```diff
@@ -59,10 +60,27 @@ func (s *SecretStoreFile) StoreSecret(username NormalizedUsername, secret []byte
 	}
 
 	final := s.userpath(username)
+
+	exists, err := FileExists(final)
+	if err != nil {
```
Comment (maxtaco) [inline go/libkb/secret_store_file.go]: need to delete the tempfile in this case (and other failures below)

File: go/libkb/secret_store_file.go [hunk 7]
```diff
@@ -59,10 +60,27 @@ func (s *SecretStoreFile) StoreSecret(username NormalizedUsername, secret []byte
...
+		return err
+	}
+
 	if err := os.Rename(f.Name(), final); err != nil {
 		return err
 	}
-	return os.Chmod(final, 0600)
+
+	if err := os.Chmod(final, 0600); err != nil {
```
Comment (maxtaco) [inline go/libkb/secret_store_file.go]: so this is a noop on non-windows?

**Comment by MarcoPolo** [comment]
(flow issue has been fixed)

File: go/libkb/secret_store_other.go [hunk 8]
```diff
@@ -6,5 +6,77 @@
...
+
+	// also log a warning (so CLI users see it)
+	g.Log.Info(pgpStorageWarningText)
+
+	// Note: a separate RPC, callable by CLI or electron, will dismiss
+	// this warning by inserting into leveldb.
+}
+
+const pgpStorageWarningText = `
+Policy change on passphrases
```
Comment (maxtaco) [inline go/libkb/secret_store_other.go]: do we get this warnings on OSX (where we do have a proper secret store)?

File: go/libkb/secret_store_other.go [hunk 9]
```diff
@@ -6,5 +6,77 @@
...
+
+We've gotten lots of feedback that it's annoying as all hell to enter a 
+Keybase passphrase after restarts and updates. The consensus is you can 
+trust a device's storage to keep a secret that's specific to that device.  
+Passphrases stink, like passed gas, and are bloody painful, like passed stones.
+
+Note, however: on this device you have a PGP private key in Keybase's local 
+keychain.  Some people want to type a passphrase to unlock their PGP key, and 
+this new policy would bypass that. If you're such a person, you can run the 
+following command to remove your PGP private key.
```
Comment (maxtaco) [inline go/libkb/secret_store_other.go]: this makes it sounds as if the key is gone forever.  We should say that we'll write it out, etc?

**Comment by maxtaco** [comment]
A few more comments...

File: go/libkb/secret_store_other.go (hunk 8 above)
Comment (patrickxb) [inline go/libkb/secret_store_other.go]: No.

File: go/libkb/secret_store_other.go (hunk 9 above)
Comment (patrickxb) [inline go/libkb/secret_store_other.go]: Text is from here: https://keybase.atlassian.net/browse/DESIGN-743

File: go/libkb/secret_store_file.go (hunk 7 above)
Comment (patrickxb) [inline go/libkb/secret_store_file.go]: yes, but can't hurt.

**Comment by maxtaco** [comment]
👍

**Pull request merged**